app/plugin_install.go

```diff
@@ -369,3 +307,88 @@ func (a *App) getBundleStorePath(id string) string {
+	pluginFile.Seek(0, 0)
+	if err := utils.ExtractTarGz(pluginFile, extractDir); err != nil {
+		return nil, extractDir, errors.New("Encountered error extracting plugin")
+	}
+
+	dir, err := ioutil.ReadDir(extractDir)
+	if err != nil {
+		return nil, extractDir, errors.Errorf("Can't read directory %s", extractDir)
+	}
+
+	if len(dir) == 1 && dir[0].IsDir() {
+		extractDir = filepath.Join(extractDir, dir[0].Name())
+	}
+
+	manifest, _, err := model.FindManifest(extractDir)
+
+	if err != nil {
+		return nil, extractDir, errors.Wrapf(err, "Can't find manifest in directory %s", extractDir)
+	}
+
+	if !plugin.IsValidId(manifest.Id) {
+		return nil, extractDir, errors.Errorf("Plugin Id must be at least %d characters, at most %d characters and match %d.", plugin.MinIdLength, plugin.MaxIdLength, http.StatusBadRequest)
+	}
+	return manifest, extractDir, nil
+}
+
+func (a *App) installExtractedPlugin(manifest *model.Manifest, fromPluginDir string, replace bool) (*model.Manifest, *model.AppError) {
+	pluginsEnvironment := a.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return nil, model.NewAppError("installExtractedPlugin", "app.plugin.disabled.app_error", nil, "", http.StatusNotImplemented)
+	}
+
+	bundles, err := pluginsEnvironment.Available()
+	if err != nil {
+		return nil, model.NewAppError("installExtractedPlugin", "app.plugin.install.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	// Check that there is no plugin with the same ID
+	for _, bundle := range bundles {
+		if bundle.Manifest != nil && bundle.Manifest.Id == manifest.Id {
+			if !replace {
+				return nil, model.NewAppError("installExtractedPlugin", "app.plugin.install_id.app_error", nil, "", http.StatusBadRequest)
+			}
+
+			if err := a.removePluginLocally(manifest.Id); err != nil {
+				return nil, model.NewAppError("installExtractedPlugin", "app.plugin.install_id_failed_remove.app_error", nil, "", http.StatusBadRequest)
+			}
+		}
+	}
+
+	pluginPath := filepath.Join(*a.Config().PluginSettings.Directory, manifest.Id)
+	err = utils.CopyDir(fromPluginDir, pluginPath)
+	if err != nil {
+		return nil, model.NewAppError("installExtractedPlugin", "app.plugin.mvdir.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	// Flag plugin locally as managed by the filestore.
+	f, err := os.Create(filepath.Join(pluginPath, managedPluginFileName))
+	if err != nil {
+		return nil, model.NewAppError("installExtractedPlugin", "app.plugin.flag_managed.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+	f.Close()
+
+	if manifest.HasWebapp() {
+		updatedManifest, err := pluginsEnvironment.UnpackWebappBundle(manifest.Id)
+		if err != nil {
+			return nil, model.NewAppError("installExtractedPlugin", "app.plugin.webapp_bundle.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		manifest = updatedManifest
+	}
+
+	// Activate plugin if it was previously activated.
```

Comment:
nit: This was probably the old comment, but it's written awkardly. 
```suggestion
	// Activate the plugin if enabled.
```